hw/dv/verilator/cpp/verilator_memutil.cc

```diff
@@ -475,62 +501,240 @@ void VerilatorMemUtil::MemWrite(const std::string &name,
+    }
+
+    uint32_t local_base = phdr.p_paddr - mem_area->addr_loc.base;
+    uint32_t local_top = lma_top - mem_area->addr_loc.base;
+
+    if (mem_area->addr_loc.size <= local_top) {
+      std::ostringstream oss;
+      oss << "Segment " << i << " has size 0x" << std::hex << phdr.p_memsz
+          << " bytes. Its LMA of 0x" << phdr.p_paddr << " is at offset 0x"
+          << local_base << " in the memory region `" << mem_area->name
+          << "', so the segment finishes at offset 0x" << local_top
+          << ", but the memory region is only 0x" << mem_area->addr_loc.size
+          << " bytes long.";
+      throw ElfError(filepath, oss.str());
+    }
+
+    // Check that the segment is aligned correctly for the memory
+    if (local_base % mem_area->width_byte) {
+      std::ostringstream oss;
+      oss << "Segment " << i << " has LMA 0x" << std::hex << phdr.p_paddr
+          << ", which starts at offset 0x" << local_base
+          << " in the memory region `" << mem_area->name
+          << "'. This offset is not aligned to the region's word width of "
+          << std::dec << 8 * mem_area->width_byte << " bits.";
+      throw ElfError(filepath, oss.str());
+    }
+
+    // Check the segment actually fits in the file
+    if (file_size < off_end) {
+      std::ostringstream oss;
+      oss << "phdr for segment " << i << " claims to end at offset 0x"
+          << std::hex << off_end - 1 << ", but the file only has size 0x"
+          << file_size << ".";
+      throw ElfError(filepath, oss.str());
+    }
+
+    const char *seg_data = file_data + phdr.p_offset;
+    WriteSegment(*mem_area, local_base, phdr.p_filesz, phdr.p_memsz,
+                 (const uint8_t *)seg_data);
   }
 }
 
-void VerilatorMemUtil::WriteVmemToMem(const svScope &scope,
-                                      const std::string &filepath) {
+const MemArea *VerilatorMemUtil::FindRegionForAddr(uint32_t lma) const {
+  // To find the memory area containing lma, use upper_bound to find the first
+  // region strictly after it, and then std::prev to step backwards. This fails
+  // if either the map is empty (obviously!) or if ub_it is already the
+  // beginning of the map.
+  if (addr_to_mem_.empty())
+    return nullptr;
+
+  auto ub_it = addr_to_mem_.upper_bound(lma);
+  if (ub_it == addr_to_mem_.begin())
+    return nullptr;
+
+  const MemArea *m = std::prev(ub_it)->second;
+
+  // Every entry in addr_to_mem_ should have a valid pointer to a MemArea with a
+  // valid location.
+  assert(m != nullptr);
+  assert(m->addr_loc.size != 0);
+
+  // What's more, the base of the location should be less than equal to lma
+  // (because it's strictly less than the smallest entry strictly greater).
+  assert(m->addr_loc.base <= lma);
+
+  // This means that we've found the right region iff the top of the region
+  // includes lma.
+  uint32_t addr_top = m->addr_loc.base + (m->addr_loc.size - 1);
+  return (lma <= addr_top) ? m : nullptr;
```

Comment:
The idea is that the `upper_bound` call finds the first region that starts strictly above `lma`. Then `std::prev(ub_it)` moves one to the left, so we should have the last region that starts at or before `lma`. The assertion is a data integrity check: we store the base of the memory area twice (once as a key in `addr_to_mem_` and once deep inside the value). It also acts as documentation: I've already edited this reply after realising the inequalities weren't coming out how I expected :-)

Anyway, the region we find starts at or before `lma`. But where does it finish? If the top of the region is also before `lma` then the load address has landed in a gap between regions (between `*m` and `ub_it->second`). In which case, we return null because there is no region that contains this address. If the top of the region is greater than `lma`, we've found a region with `base <= lma < top`. Success!